crates/ra_assists/src/add_new.rs

```diff
@@ -0,0 +1,354 @@
+    let impl_node = ctx.node_at_offset::<ast::ImplBlock>()?;
+    let item_list = impl_node.item_list()?;
+    let items: Vec<_> = item_list.impl_items().collect();
+
+    // Don't attempt to create a `new` function if one already exists.
+    let fn_name = |item: &ast::ImplItem| match item.kind() {
+        ast::ImplItemKind::FnDef(def) => def.name(),
+        _ => None,
+    };
+    if items.iter().filter_map(fn_name).any(|n| n.text() == "new") {
+        return None;
+    }
+
+    // Find the type that this impl block is for, returning None if that type is not a struct.
+    let struct_def = {
+        let file_id = ctx.frange.file_id;
+        let position = FilePosition { file_id, offset: impl_node.syntax().text_range().start() };
+        let analyzer = hir::SourceAnalyzer::new(ctx.db, position.file_id, impl_node.syntax(), None);
+        resolve_target_struct_def(ctx.db, &analyzer, &impl_node)?
+    };
+
+    let struct_name = struct_def.name()?;
+    let mut children = struct_def.syntax().children();
+    let field_def_list = children.find_map(ast::NamedFieldDefList::cast)?;
+
+    ctx.add_action(AssistId("add_new"), "add new", |edit| {
+        // We'll overwrite the contents of the impl body from its opening curly brace to its
+        // first non-whitespace token, which for empty bodies will be their closing brace.
+        let l_paren = item_list.syntax().children_with_tokens().find(|it| it.kind() == T!['{']);
+        if l_paren.is_none() {
+            return;
+        }
+        let replace_start = l_paren.unwrap().text_range().start() + TextUnit::from_usize(1);
+        let first_element_inside_body = item_list
+            .syntax()
+            .children_with_tokens()
+            .skip_while(|it| it.kind() != T!['{'])
+            .skip_while(|it| it.kind() == T!['{'])
+            .find(|it| !it.kind().is_whitespace());
+        if first_element_inside_body.is_none() {
+            return;
+        }
+        let replace_end = first_element_inside_body.clone().unwrap().text_range().start();
+        let replace_range = TextRange::from_to(replace_start, replace_end);
+
+        // Construct the `new` function. Note that the indentation here is hard-coded.
+        // See https://github.com/rust-analyzer/rust-analyzer/issues/1645
+        let mut buf = String::new();
+        buf.push_str("\n    fn new(");
+        let fields: Vec<_> = field_def_list
+            .fields()
+            .filter_map(|f| Some((f.name()?.syntax().text(), f.ascribed_type()?.syntax().text())))
+            .collect();
+        let mut first = true;
+        for field in &fields {
+            if !first {
+                buf.push_str(", ");
+            }
+            buf.push_str(&format!("{}: {}", field.0, field.1));
+            first = false;
+        }
+        buf.push_str(") -> Self {\n");
+        buf.push_str(&format!("        {} {{", struct_name.syntax().text()));
```

Comment:
Can we use `Self` type instead of structure name here? Wouldn’t it be better for future refactoring, for instance, in a case when structure type is renamed? This renaming would require fewer changes.

Reply:
Good idea. Done.